### PR TITLE
Adds a nonlinear preconditioned grain growth model

### LINKED ISSUE #21940: Add nonlinear preconditioning grain growth model
## Reason
The existing phase field grain growth model requires at least three elements across the diffuse interface. Nonlinear preconditioning  can reduce that and thus reduce the computational cost of modeling grain growth.

## Design
This will introduce a change of variable so that the nonlinear variable is linear across the interface rather than following a tanh function.

## Impact
This will not change the existing grain growth model, but just add the new nonlinear preconditioned model.

### PR DESCRIPTION
**Pull request by tonkmr**
## Reason
Adds nonlinear preconditioning to the grain growth model, to reduce the required mesh size. closes #21940.

## Design
It adds:
- New kernels
- New ics
- A new material class
- An action that automates it all
- tests
- All documentation

## Impact
Does not impact any existing models, but enables more efficient grain growth simulations

### CODE REVIEW

**Comment by tonkmr** [comment]
@permcody I added a GrainTracker test for the new functionality and it is failing in some cases, but not all. It passes fine on my computer. It looks like it is failing because on some systems GrainTracker remaps to a different empty variable than on other systems. This isn't necessarily wrong, its just different. Do you have any recommendations? I could just use fewer order parameters so it only has one option.

**Comment by tonkmr** [comment]
@dschwen a good place to start with this PR would be the new documentation page Nonlin_Precond_Grain_Growth.md. It summarizes the approach and all of the new objects that I have created to implement it.

**Comment by tonkmr** [comment]
@dschwen Check SQA is failing on the Documentation test. It says `/AuxKernels/NonlinPrecondAux` is missing a markdown file, but it isn't. I have added that file. It seems to be looking for it in `framework` instead of `modules` but I'm not sure why.

File: modules/phase_field/src/auxkernels/NonlinPrecondAux.C
Comment (GiudGiud) [inline modules/phase_field/src/auxkernels/NonlinPrecondAux.C:12]: this should be phase field app not MooseApp

this is why SQA is failing

Comment (GiudGiud) [reply modules/phase_field/src/auxkernels/NonlinPrecondAux.C:12]: (eg because it sees mooseApp, it looks in framework/doc instead of modules...doc

**Comment by tonkmr** [comment]
@laagesen This is the grain growth nonlinear preconditioning model that I have been working on. Please review.

File: modules/phase_field/include/actions/GrainGrowthNonlinPrecondAction.h
Comment (dschwen) [inline modules/phase_field/include/actions/GrainGrowthNonlinPrecondAction.h:19]: This looks similar enough to `GrainGrowthAction` that some refactoring would be in order.

File: modules/phase_field/include/kernels/ACGrGrPoly.h
Comment (dschwen) [inline modules/phase_field/include/kernels/ACGrGrPoly.h:30]:
```suggestion
  virtual Real assignThisOP();
  virtual std::vector<Real> assignOtherOPs();
```

File: modules/phase_field/src/actions/PolycrystalColoringICAction.C
Comment (dschwen) [inline modules/phase_field/src/actions/PolycrystalColoringICAction.C:23]:
```suggestion
      "Action to create ICs for polycrystal variables defined using a Voronoi tessellation");
```

File: modules/phase_field/src/actions/PolycrystalColoringICAction.C
Comment (dschwen) [inline modules/phase_field/src/actions/PolycrystalColoringICAction.C:34]: Rather than the all cap comment how about putting both parameters into a `Nonlinear Preconditioning` parameter group?

File: modules/phase_field/src/actions/PolycrystalColoringICAction.C
Comment (dschwen) [inline modules/phase_field/src/actions/PolycrystalColoringICAction.C:44]: I'd love to see all instances of "nonlin precond" expanded to "nonlinear preconditioning"

File: modules/phase_field/src/auxkernels/NonlinPrecondAux.C
Comment (dschwen) [inline modules/phase_field/src/auxkernels/NonlinPrecondAux.C:33]:
```suggestion
  return (1.0 + std::tanh(_phi[_qp] / std::sqrt(2.0))) / 2.0;
```

File: modules/phase_field/src/kernels/ACGrGrPolyNonlinPrecond.C
Comment (dschwen) [inline modules/phase_field/src/kernels/ACGrGrPolyNonlinPrecond.C:80]:
```suggestion
    case Residual:
      return ACGrGrPoly::computeDFDOP(type);

    case Jacobian:
      return ACGrGrPoly::computeDFDOP(type) * _dopdphi[_qp];
```

File: modules/phase_field/src/utils/ExpressionBuilder.C
Comment (dschwen) [inline modules/phase_field/src/utils/ExpressionBuilder.C:260]: :+1:

File: modules/phase_field/src/actions/PolycrystalColoringICAction.C
Comment (tonkmr) [reply modules/phase_field/src/actions/PolycrystalColoringICAction.C:44]: @dschwen do you mean just variable names like here or everywhere, including the file names?

File: modules/phase_field/src/actions/PolycrystalColoringICAction.C
Comment (tonkmr) [reply modules/phase_field/src/actions/PolycrystalColoringICAction.C:34]: @dschwen I can't remember the syntax for parameter groups. Can you point me at an example?

Comment (GiudGiud) [reply modules/phase_field/src/actions/PolycrystalColoringICAction.C:34]: params.addParamNamesToGroup("shift scale", "group name");

File: modules/phase_field/src/actions/PolycrystalColoringICAction.C
Comment (tonkmr) [reply modules/phase_field/src/actions/PolycrystalColoringICAction.C:44]: @dschwen @laagesen We talked in person about changing the name of this because "nonlinear preconditioning" might make users think this is related to preconditioning in PETSc. @laagesen suggested "change-of-variable" but I think that might be too generic. Here are some other options I have thought of:

- Linear Interface
- Linear Substitution
- Linear Variable Substitution

Which do you prefer?

File: modules/phase_field/include/actions/GrainGrowthNonlinPrecondAction.h
Comment (tonkmr) [reply modules/phase_field/include/actions/GrainGrowthNonlinPrecondAction.h]: I addressed this comment and refactored GrainGrowthAction and now this action inherits from it.

File: modules/phase_field/include/kernels/ACGrGrPoly.h
Comment (tonkmr) [reply modules/phase_field/include/kernels/ACGrGrPoly.h]: @dschwen I have addressed this comment

File: modules/phase_field/src/actions/PolycrystalColoringICAction.C
Comment (tonkmr) [reply modules/phase_field/src/actions/PolycrystalColoringICAction.C]: @dschwen I have addressed this comment

File: modules/phase_field/src/actions/PolycrystalColoringICAction.C
Comment (tonkmr) [reply modules/phase_field/src/actions/PolycrystalColoringICAction.C]: I have addressed this comment

File: modules/phase_field/src/auxkernels/NonlinPrecondAux.C
Comment (tonkmr) [reply modules/phase_field/src/auxkernels/NonlinPrecondAux.C]: I have addressed this comment

File: modules/phase_field/src/kernels/ACGrGrPolyNonlinPrecond.C
Comment (tonkmr) [reply modules/phase_field/src/kernels/ACGrGrPolyNonlinPrecond.C]: I have addressed this comment

File: modules/phase_field/src/actions/PolycrystalColoringICAction.C
Comment (dschwen) [reply modules/phase_field/src/actions/PolycrystalColoringICAction.C:45]: The substitution is not linear, but it _linearizes_ the order parameters. `LinearizedInterface` would be my sugegstion.

Comment (tonkmr) [reply modules/phase_field/src/actions/PolycrystalColoringICAction.C:45]: I like LinearizedInterface

Comment (laagesen) [reply modules/phase_field/src/actions/PolycrystalColoringICAction.C:45]: I also like `LinearizedInterface`.

**Comment by tonkmr** [comment]
@GiudGiud @dschwen  Two internal checks are failing and I am not able to see the results. Can you tell me what the errors are?

**Comment by GiudGiud** [comment]
Seems marmot does not build anymore
@dschwen

**Comment by tonkmr** [comment]
I assume that there is another action in MARMOT that inherits from GrainGrowthAction, but I don't have access to MARMOT to fix it.

**Comment by laagesen** [comment]
@GiudGiud I tried to look at the internal app tests and cannot see them, not sure if this might be related to the switch over to INL Github, I will contact you via Teams about this...

@tonkmr Did you lose access to Marmot due to the switch from Gitlab to Github, or did it lapse before that? Do you want to get access restored?

**Comment by loganharbour** [comment]
@laagesen the failure is in marmot. If you're signed in via github.com on civet, you should be able to see the results now.

**Comment by dschwen** [comment]
Yeah, it's my `SwellingGrainGrowthAction` :-/

**Comment by dschwen** [comment]
I'm working on a Marmot PR to fix this

**Comment by laagesen** [comment]
thanks @dschwen - with Logan's help I can see the internal tests now

File: modules/phase_field/include/actions/GrainGrowthAction.h
Comment (dschwen) [inline modules/phase_field/include/actions/GrainGrowthAction.h:29]: Why is this being removed?

File: modules/phase_field/src/actions/GrainGrowthAction.C
Comment (dschwen) [inline modules/phase_field/src/actions/GrainGrowthAction.C:127]: undo please

File: modules/phase_field/src/actions/GrainGrowthAction.C
Comment (dschwen) [inline modules/phase_field/src/actions/GrainGrowthAction.C:255]: Undo this please, this is what breaks Marmot!

File: modules/phase_field/include/actions/GrainGrowthAction.h
Comment (dschwen) [reply modules/phase_field/include/actions/GrainGrowthAction.h:29]: Hmm, and while you're at it would you mind making this virtual. I'm puzzled how this ever worked without being virtual...

Comment (dschwen) [reply modules/phase_field/include/actions/GrainGrowthAction.h:29]: actually scratch that

File: modules/phase_field/src/actions/GrainGrowthAction.C
Comment (dschwen) [reply modules/phase_field/src/actions/GrainGrowthAction.C:127]: hold off on that

File: modules/phase_field/src/actions/GrainGrowthAction.C
Comment (dschwen) [reply modules/phase_field/src/actions/GrainGrowthAction.C:255]: I can fix Marmot instead. This function isn't really needed.

**Comment by dschwen** [comment]
Marmot PR is ready.

**Comment by tonkmr** [comment]
Thanks! I still need to rename everything, but then it will be ready to go.

Mike

--------------
Michael Tonks
Associate Chair and Alumni Professor
Materials Science and Engineering
University of Florida

100D Rhines Hall, Gainesville, FL 32611
(352) 846-3779
***@***.***


On Jul 10, 2023, at 5:19 PM, Daniel Schwen ***@***.***> wrote:

[External Email]


Marmot PR is ready.

**Comment by tonkmr** [comment]
@GiudGiud you mentioned on Slack that since I have a documentation page doc/source/actions/GrainGrowthLinearizedInterfaceAction.md, I do not need to create doc/syntax/Modules/PhaseField/GrainGrowthLinearizedInterface/index.md. However, the Documentation tests are failing with the error: "Unable to locate a page that ends with the name 'syntax/Modules/PhaseField/GrainGrowthLinearizedInterface/index.md". Any idea on what might be causing this? Thanks!

**Comment by GiudGiud** [comment]
Seems I was wrong. You need a syntax documentation page because you introduced a new syntax

**Comment by tonkmr** [comment]
@dschwen @laagesen @GiudGiud I have addressed all of the comments, so this is ready to review again. I would like to wait and transition the LinearizedInterfaceMaterial away from ExpressionBuilder in a new PR.

**Comment by GiudGiud** [comment]
I'll let @dschwen decide on the next steps here

**Review by laagesen — APPROVED** [review]

**Review by dschwen — APPROVED** [review]

File: modules/phase_field/test/tests/actions/gold/grain_growth_out.e
Comment (dschwen) [reply modules/phase_field/test/tests/actions/gold/grain_growth_out.e:1]: Looks like those are not used :-/

**Review by dschwen — DISMISSED** [review]
Please remove unused files.

File: modules/phase_field/test/tests/actions/gold/grain_growth_out.e
Comment (GiudGiud) [reply modules/phase_field/test/tests/actions/gold/grain_growth_out.e:1]: I see the test exists already,.

are these regolds? It looks like a new file on git

Comment (GiudGiud) [reply modules/phase_field/test/tests/actions/gold/grain_growth_out.e:1]: I get this diff with exodiff:
```
  ==============================================================
  NOTE: All node and element ids are reported as global ids.

Sideset Distribution Factors:
  --------- Time step 1, 0.0000000e+00 ~ 0.0000000e+00, rel diff:  0.00000e+00 ---------
Global variables:
Nodal variables:
   bnds  rel diff:  0.0000000e+00 ~  1.0000000e+00 = 1.00000e+00 (node 1)
  --------- Time step 2, 8.0000000e+01 ~ 8.0000000e+01, rel diff:  0.00000e+00 ---------
Global variables:
Nodal variables:
  --------- Time step 3, 1.6000000e+02 ~ 1.6000000e+02, rel diff:  0.00000e+00 ---------
Global variables:
Nodal variables:
  --------- Time step 4, 2.4000000e+02 ~ 2.4000000e+02, rel diff:  0.00000e+00 ---------
Global variables:
Nodal variables:
  --------- Time step 5, 3.2000000e+02 ~ 3.2000000e+02, rel diff:  0.00000e+00 ---------
Global variables:
Nodal variables:
  --------- Time step 6, 4.0000000e+02 ~ 4.0000000e+02, rel diff:  0.00000e+00 ---------
Global variables:
Nodal variables:

exodiff: Files are different
```

Comment (tonkmr) [reply modules/phase_field/test/tests/actions/gold/grain_growth_out.e:1]: I changed the action so that the BndsCalcAux runs on initial, when it didn't previously. So, I had to re-gold the tests. The input files are the same, but the action is different, so the files had to change.

Comment (GiudGiud) [reply modules/phase_field/test/tests/actions/gold/grain_growth_out.e:1]: alright thanks for double checking

**Review by GiudGiud — APPROVED** [review]

**Pull request merged**
